lab_4/main.py

```diff
@@ -3,45 +3,144 @@
+            raise ValueError
+
+        if word not in self.storage:
+            self.storage[word] = self.id
+            self.id += 1
+
+        return self.storage[word]
 
     def get_id(self, word: str) -> int:
-        pass
+
+        if not isinstance(word, str):
+            raise ValueError
+
+        if word not in self.storage:
+            raise KeyError
+
+        return self.storage[word]
 
     def get_word(self, word_id: int) -> str:
-        pass
+
+        if not isinstance(word_id, int):
+            raise ValueError
+
+        for key, word_id_another in self.storage.items():
+            if word_id == word_id_another:
+                return key
+
+        raise KeyError
 
     def update(self, corpus: tuple):
-        pass
+
+        if not isinstance(corpus, tuple):
+            raise ValueError
+
+        for word in corpus:
+            self._put_word(word)
 
 
 def encode_text(storage: WordStorage, text: tuple) -> tuple:
-    pass
+
+    if not isinstance(storage, WordStorage) or not isinstance(text, tuple):
+        raise ValueError
+
+    encoded_corpus = []
+    for word in text:
+        word_id = storage.get_id(word)
+        encoded_corpus.append(word_id)
+
+    return tuple(encoded_corpus)
 
 
 class NGramTextGenerator:
     def __init__(self, word_storage: WordStorage, n_gram_trie: NGramTrie):
-        pass
+        self._word_storage = word_storage
+        self._n_gram_trie = n_gram_trie
 
     def _generate_next_word(self, context: tuple) -> int:
-        pass
+
+        if not isinstance(context, tuple) or len(context) + 1 != self._n_gram_trie.size:
+            raise ValueError
+
+        max_frequency = 0
+        full_context = ()
+
+        for key, frequency in self._n_gram_trie.n_gram_frequencies.items():
+            if key[:len(context)] == context and frequency > max_frequency:
+                max_frequency = frequency
+                full_context = key
+
+        if not full_context:
+            return max(self._n_gram_trie.uni_grams, key=self._n_gram_trie.uni_grams.get)[0]
+
+        return full_context[-1]
 
     def _generate_sentence(self, context: tuple) -> tuple:
-        pass
+
+        if not isinstance(context, tuple):
+            raise ValueError
+
+        sentence_generated = list(context)
+        for i in range(20):
+            sentence_generated.append(self._generate_next_word(tuple(sentence_generated[-(len(context)):])))
+            if sentence_generated[-1] == self._word_storage.storage['<END>']:
+                break
+
+        else:
+            sentence_generated.append(self._word_storage.storage['<END>'])
+        return tuple(sentence_generated)
 
     def generate_text(self, context: tuple, number_of_sentences: int) -> tuple:
-        pass
+
+        if not isinstance(context, tuple) or not isinstance(number_of_sentences, int):
+            raise ValueError
+
+        text_generated = []
+
+        for i in range(number_of_sentences):
```

Comment:
If you do not intend to use the var from `for`-loop, it is commonly named as `_`.